mountpoint-s3/src/inode.rs

```diff
@@ -360,6 +360,68 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Unlink the entry described by `parent_ino` and `name`.
+    ///
+    /// If the entry exists, delete it from S3 and the superblock.
+    pub async fn unlink<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        parent_ino: InodeNo,
+        name: &OsStr,
+    ) -> Result<(), InodeError> {
+        let parent = self.inner.get(parent_ino)?;
+        let LookedUp { inode, .. } = self.lookup(client, parent_ino, name).await?;
+
+        if inode.kind() == InodeKind::Directory {
+            return Err(InodeError::IsDirectory(inode.ino()));
+        }
+
+        let write_status = {
+            let inode_state = inode.inner.sync.read().unwrap();
+            inode_state.write_status
+        };
+
+        match write_status {
+            WriteStatus::LocalUnopened | WriteStatus::LocalOpen => {
+                // In the future, we may permit `unlink` and cancel any in-flight uploads.
+                debug!("unlink called on local file, unlink not supported until write is complete");
+                return Err(InodeError::UnlinkNotPermittedWhileWriting(inode.ino()));
+            }
+            WriteStatus::Remote => {
+                debug!("unlink called on remote file, will delete from S3 and unlink from parent");
+                let (bucket, s3_key) = (self.inner.bucket.as_str(), inode.full_key());
+                let delete_obj_result = client.delete_object(bucket, s3_key).await;
+
+                match delete_obj_result {
+                    Ok(_res) => debug!("object with key {s3_key} successfully deleted, if it existed"),
+                    Err(e) => {
+                        error!(error=?e, "unlink failed when trying to perform S3 DeleteObject call, not unlinking from parent inode");
+                        Err(InodeError::ClientError(e.into()))?;
+                    }
+                };
+            }
+        }
+
+        let mut parent_state = parent.inner.sync.write().unwrap();
+        match &mut parent_state.kind_data {
+            InodeKindData::File { .. } => {
+                debug_assert!(false, "inodes never change kind");
+                return Err(InodeError::NotADirectory(parent.ino()));
+            }
+            InodeKindData::Directory {
+                children,
+                writing_children,
+            } => {
+                writing_children.remove(&inode.ino());
+                if children.remove(inode.name()).is_none() {
+                    debug!("child was not present, another operation may have occurred in parallel");
```

Comment:
I think we want to arrange the locking/state such that this state isn't possible. Maybe what we need is to record in the inode itself that a DeleteObject is in flight for the inode.

Reply:
I had similar thoughts, but I'm not sure what we should expect of `unlink` operations. 

- Should we try to block subsequent `unlink` until the first is complete? My thinking there is that we'd have to add something like `unlink_mutex: Mutex<()>` which can let us serialize `unlink` operations.
- Should we return an error... but what error?
- Should we do something like @passaro suggests which is to continue with the `unlink`, and only remove the child if the `Inode` or `InodeNo` is equal to the one we saw at the beginning of the `unlink` operation?

Reply:
I'm leaning towards serializing the unlinks, since `unlink` would be atomic for POSIX FS (right?). Thus, we could then return `ENOENT` ("error, no entry found") for subsequent `unlink` requests.

Thoughts?